File: 600/637 - Average of Levels in Binary Tree.java
import java.util.ArrayList;
import java.util.Queue;
import java.util.LinkedList;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Double> averageOfLevels(TreeNode root) {
        int len, i;
        double sum;
        TreeNode curr;
        List<Double> ret = new ArrayList<Double>();
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        
        if (root == null)
            return ret;
        queue.offer(root);
        
        while (! queue.isEmpty()) {
            sum = 0.0;
            len = queue.size();
            
            for (i = 0; i < len; ++i) {
                curr = queue.poll();
                if (curr.left != null)
                    queue.offer(curr.left);
                if (curr.right != null)
                    queue.offer(curr.right);
                
                sum += curr.val;
            }
            
            ret.add(sum / (1.0 * len));
        }
        
        return ret;
    }
}

File: csharp/637.average-of-levels-in-binary-tree.cs
/*
 * @lc app=leetcode id=637 lang=csharp
 *
 * [637] Average of Levels in Binary Tree
 */

using System.Collections.Generic;

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public IList<double> AverageOfLevels(TreeNode root) {
        IList<double> ret = new List<double>();
        if (root == null) {
            return ret;
        }
        Queue<TreeNode> q = new Queue<TreeNode>();
        q.Enqueue(root);
        while (q.Count > 0) {
            int x = q.Count;
            double acc = 0;
            for (int i = 0; i < x; ++i) {
                var curr = q.Dequeue();
                acc += curr.val;
                if (curr.left != null) {
                    q.Enqueue(curr.left);
                }
                if (curr.right != null) {
                    q.Enqueue(curr.right);
                }
            }
            ret.Add(acc / (double) x);
        }
        return ret;
    }
}
// @lc code=end

